add R and cl documentation to accompany existing python documentation in Bulk Processing (#519)
Sage-Bionetworks/synapseDocs: articles/uploading_in_bulk.md

---
title: Uploading and Downloading Data in Bulk
layout: article
excerpt: Learn about uploading files with annotations and provenance in bulk.
category: howto
---

# Uploading and Downloading Data in Bulk

> Uploading a large number of files can be tedious when uploading one at a time on the web, especially if you want to set [annotations](http://docs.synapse.org/articles/annotation_and_query.html) and [provenance](http://docs.synapse.org/articles/provenance.html). The command line, Python client and R client have a convenient function for bulk upload and download called `sync`. The uploads require a tab delimited *manifest* where each file to be uploaded and it's annotations are specified as a row in the manifest file. In this article we will cover how to: 

Reviewer: ```suggestion
Working with a large number of files on the web can be tedious, especially if you want to download, upload, or set [annotations](http://docs.synapse.org/articles/annotation_and_query.html) and [provenance](http://docs.synapse.org/articles/provenance.html). The command line, Python client and R client have convenience functions for bulk upload and download. Uploading require a tab delimited *manifest* where each file to be uploaded and, optionally, annotations to be applied, are specified as a row in the file. Downloading in bulk requires identifying a container (`Folder`, `Project`, `Table`, or `View`) that contains the files of interest. In this article we will cover how to: 
```

	
* create a manifest 
* upload the files in bulk
* modify files in bulk using a manifest
* download the files in bulk

<br/>

## Uploading Data in Bulk

### Creating a Manifest

Files to be uploaded are specified in a tab separated (`.tsv`) manifest. The manifest has columns that contain information about each file to be uploaded along with annotations that will be associated with the file in Synapse.

> The required columns in the manifest are `path`, the directory of the file to be uploaded, and `parent`, the synId of the folder. It is optional to link code repositories via the `executed` column. Here is an example manifest that uploads a single file:

Reviewer: ```suggestion
The required columns in the manifest are:

 - `path`: the directory of the file to be uploaded
 -  `parent`: the Synapse ID of the Folder to upload to
 
 It is optional to link provenance - the `used` column can indicate files that were used to create the one being uploaded, and the `executed` column can indicate code (in Synapse or on the web) that was used to generate the file. Here is an example manifest that uploads a single file:
```

<br/>

> The above manifest describes a "file.csv" that will be uploaded to the Synapse folder `syn123` and named "Tardar Sauce". The manifest describes the provenance of the file indicating that it was generated using code deposited in GitHub (https://github.com/your/code/repo) from the data in `syn654`. Additionally, the file has been annotated with; `emotion:grumpy` and `species:cat`. Additonal annotations are associated with the file by adding column variables (e.g. assay, fileFormat, cellType, etc.).

Reviewer: ```suggestion
The above manifest describes a "file.csv" that will be uploaded to the Synapse folder `syn123` and named "Tardar Sauce". The manifest describes the provenance of the file indicating that it was generated using code deposited in GitHub (https://github.com/your/code/repo) from the data in `syn654`. Additionally, the file has been annotated with `emotion: grumpy` and `species: cat`. Additional annotations could be associated with the file by adding more columns.
```
